roachtest: interface for goroutines

herkolategan opened the pull request:
Previously, roachtests would use bare goroutines `go func...` to start tasks
during a test to perform various background functions. This is problematic due
to the nature of roachtest, which runs multiple tests in parallel using workers.
If one such bare goroutine panics the whole test suite is interrupted and the
process exits.

This change is the first step and introduces an interface that both the
traditional and mixedversion test frameworks can utilise. It intends to target
both regular and MVT tests by supplying the test with the interface either
through a helper or the test interface.

In addition to providing the `Go` function, there are additional options that
can be passed. The test framework will generally supply the panic and error
handler, but users can supply additional options such as naming the task, or
passing a custom logger. This makes logging more useful to determine which
task caused a panic or encountered an error.

Informs: #118214

Epic: None
Release note: None

herkolategan committed:
roachtest: interface for goroutines

Previously, roachtests would use bare goroutines `go func...` to start tasks
during a test to perform various background functions. This is problematic due
to the nature of roachtest, which runs multiple tests in parallel using workers.
If one such bare goroutine panics the whole test suite is interrupted and the
process exits.

This change is the first step to address goroutine concerns in roachtest. It
introduces an interface that both the traditional and mixedversion test
frameworks can utilise. The test framework will supply tests with the interface
either through a helper or the test interface.

In addition to providing the `Go` function, there are additional options that
can be passed. The test framework will generally supply the panic and error
handler, but users can supply additional options such as naming the task, or
passing a custom logger. This makes logging more useful to determine which task
encountered a panic or an error.

Informs: #118214

Epic: None
Release note: None

herkolategan committed:
roachtest: task manager

This change provides a task manager implementation that will drive the logic
running behind the Tasker interface. The Manager interface and implementation is
meant to be used by the test framework(s), while the Tasker interface is
supplied to tests. The framework will do additional logic such as terminating
tasks at the end of a test, logging uncaught errors or issues and providing the
appropriate panic and error handlers.

Informs: #118214

Epic: None
Release note: None

herkolategan committed:
roachtest: refactor mixedversion background runner

Previously, the mixedversion framework had its own implementation for managing
background tasks. This implementation performed well, but in its current state
it is difficult to share with tests that do not require the mixedversion
framework.

This change replace the mixedversion framework background runner with the newly
implemented task manager. The implementation details are relatively similar
making it close to a drop-in replacement. Additionally, some utilities relating
to tasks have been replaced by the task package.

Informs: #118214

Epic: None
Release note: None

herkolategan committed:
roachtest: replace bare goroutine in mixedversion test

This change now utilises the Tasker interface supplied by the mixedversion
helper to start a task. In essence a background function can also be utilised,
but for low impact go routines the utility provides a quick way of starting a
task.

Informs: #118214

Epic: None
Release note: None

herkolategan committed:
roachtest: add task manager for test implementation

Previously, the Test interface and implementation did not provide any way to
manage goroutines during a test. This change adds the relevant code to bootstrap
the task manager during a test and do cleanup. The Test interfaces now exposes
the Tasker interface for roachtests that start goroutines (tasks).

As with the mixedversion framework, when a task fails (and is not handled
explicitly) in a test it will result in a test failure.

Informs: #118214

Epic: None
Release note: None

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/helper.go:
```diff
@@ -224,29 +225,42 @@ func (h *Helper) ExecWithGateway(
-	name string, fn func(context.Context, *logger.Logger) error,
-) context.CancelFunc {
-	return h.runner.background.Start(name, func(ctx context.Context) error {
+// GoWithCancel implements the Tasker interface.
```
I think I understand why we need to both implement the Tasker interface and have a `Background` function: the former is used by backup helpers shared between mvt and non mvt tests, while the latter is what existed before and formats the task name for us.

A small nit would be that it could be confusing to a mvt writer which of the three to use. Perhaps a comment guiding users towards `Background` would be good?

herkolategan replied:
I agree, this thought crossed my mind, and I planned to do a follow-up PR. Just wanted to avoid doing too much in one PR. But I'll add a commit to refactor `background` helper functions -> `tasker`.

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1287,6 +1289,26 @@ func (r *testRunner) runTest(
 
+	// Monitor the task manager for completed events, or failure events and log
+	// them. A failure will call t.Errorf which cancels the test's context.
+	go func() {
```
Should we add a mocked roachtest in `test_test.go` to test this logic out? We already have something similar to test timeouts, so I think it should be doable.

herkolategan replied:
Good call! I added a test, and it caught a bug! Was using the wrong context for the task manager. Test covers that scenario now as well.

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1563,6 +1585,10 @@ func (r *testRunner) teardownTest(
 
+	// Terminate tasks to ensure that any stray tasks are cleaned up.
+	t.L().Printf("terminating tasks")
+	t.taskManager.Terminate(t.L())
```
Is there a reason we only want to terminate task on successful tests? If not, should we move this to a defer at the top? Otherwise it's possible we return before we reach here:

https://github.com/cockroachdb/cockroach/blob/9ec5ed022888460e8e808c18a7a96ad742882512/pkg/cmd/roachtest/test_runner.go#L1574

herkolategan replied:
Nice catch, didn't notice the short-circuit above! I'll amend.

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1563,6 +1585,10 @@ func (r *testRunner) teardownTest(
 		getGoCoverArtifacts(ctx, c, t)
 	}
 
+	// Terminate tasks to ensure that any stray tasks are cleaned up.
```
Question: if the test is successful, do we expect any tasks to not be cancelled already? If not maybe we should log when `Terminate` cleans up a task. I'm thinking it could be a warning the test writer maybe isn't properly cancelling their goroutines?

herkolategan replied:
Yes, I guess it depends on how the test implements it. A warning might be worthwhile, I'll see if I can add one.

herkolategan replied:
This is a little more involved than I anticipated since there is no easy way to tell if a group is still busy or have active tasks, so this would require more involved task tracking. Currently it has a mechanism that will complain about tasks that do not terminate within 5 minutes, whether the test failed or not. I think the idea to log 'leaky' tasks is still worthwhile, but maybe in a follow-up PR?

DarrylWong reviewed (commented):
Left a few small comments but overall this is looking good! Should we kick off a test run of the mixed version tests, since they appear to be the most affected by this change?

Question: are we going to eventually disallow bare goroutines in general in favor of tasks instead?

herkolategan commented:
> Left a few small comments but overall this is looking good! Should we kick off a test run of the mixed version tests, since they appear to be the most affected by this change?
> 
> Question: are we going to eventually disallow bare goroutines in general in favor of tasks instead?

Thanks for the review, some great suggestions. I'll make the changes and do a run of the mixed version tests.

herkolategan committed:
roachtest: refactor mixedversion background functions

Previously, mixedversion tests called `background` on the helper passed with the
test. With the introduction of the `Tasker` interface these calls have been
refactored to use the new naming and options for passing a task name.

Informs: #118214

Epic: None
Release note: None

DarrylWong commented on pkg/cmd/roachtest/tests/multitenant_upgrade.go:
```diff
@@ -156,13 +157,10 @@ func runMultitenantUpgrade(ctx context.Context, t test.Test, c cluster.Cluster)
-					return fn(ctx, l, tenant)
-				},
-			)
+			h.Go(func(ctx context.Context, l *logger.Logger) error {
```
Nice! I think I like this more than before, where `Background` referred to either a specific stage and also the goroutine wrapper provided.

DarrylWong commented on pkg/cmd/roachtest/test_test.go:
```diff
@@ -294,6 +295,7 @@ func (b *syncedBuffer) String() string {
 }
 
 func TestRunnerTestTimeout(t *testing.T) {
+
```
nit: extra blank line :)

DarrylWong reviewed (approved):
Nice, thanks for adding the tests! This LGTM assuming the MVTs look good.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/task/manager.go:
```diff
@@ -0,0 +1,143 @@
+		event := Event{
+			Name:           opt.Name,
+			Err:            err,
+			ExpectedCancel: err != nil && IsContextCanceled(groupCtx) && expectedContextCancellation,
```
Just wondering why `ExpectedCancel` is conditional on `err != nil`. What if the task is cancelled _without_ an error? Also, doesn't `expectedContextCancellation == true` imply `IsContextCanceled` since the former is set iff context is cancelled?

herkolategan replied:
I renamed this during refactoring from the background runner. It was originally named `TriggeredByTest`. I think I understand what Renato's intentions was here. He did not want the test framework to report an error if it was indeed caused by a cancellation. Hence `err != nil` and we are sure the context was cancelled and expected to be cancelled. The latter two checks ensure, we don't have a `expectedContextCancellation`, but in-between setting that and the task running we did not encounter a different error.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/task/manager.go:
```diff
@@ -0,0 +1,143 @@
+	})
+
+	taskCancelFn := func() {
+		expectedContextCancellation = true
```
This write may happen from a different thread than the read (in `ExpectedCancel`); so, technically this should be an atomic var.

herkolategan replied:
Yep, this could probably also trigger the race detector. Updated.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/task/manager.go:
```diff
@@ -0,0 +1,143 @@
+		case m.events <- event:
+			// exit goroutine
+		case <-m.ctx.Done():
+			// Parent context already finished, exit goroutine.
```
Recall that select may execute cases out of order, when both conditions are ready. So, if the context was cancelled, we may end up missing the corresponding event.

herkolategan replied:
> Recall that select may execute cases out of order

definitely!

This logic above is intentional, and also stems from the original logic not to forward events if the parent context is already done / cancelled.  We likely already have an error like "timeout", or something else that cancelled the context, meaning errors from tasks don't have much meaning after this point. Or at least I figure that was why this logic was structured like this in the mixedversion test background runner. Thoughts?

herkolategan replied:
Also answered this in part below, since the way panics are handled relies on this logic to prevent duplicate logging.

srosenberg replied:
> We likely already have an error like "timeout", or something else that cancelled the context, meaning errors from tasks don't have much meaning after this point. Or at least I figure that was why this logic was structured like this in the mixedversion test background runner. 

Makes sense; I sorta assumed that was the reasoning behind it. However, an explicit commit could have clarified why it's okay to drop the event in this case.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/task/manager.go:
```diff
@@ -0,0 +1,143 @@
+		_ = m.group.Wait()
+	}()
+
+	WaitForChannel(doneCh, "tasks", l)
```
The above goroutine may leak when `WaitForChannel` returns (i.e., exceeds the 5-minute timeout). It would also violate the contract of `Terminate`, since according to the comment, all task functions are expected to have returned.

herkolategan replied:
There is not much we can do about leaking, except log that we could not terminate all tasks. But the comment is misleading in that regard, I'll update it to reflect that this is best effort.

srosenberg commented on pkg/cmd/roachtest/test_impl.go:
```diff
@@ -648,6 +653,22 @@ func (t *testImpl) IsBuildVersion(minVersion string) bool {
+
+// Go is like GoWithCancel but without a cancel function.
+func (t *testImpl) Go(fn task.Func, opts ...task.Option) {
+	_ = t.GoWithCancel(fn, task.OptionList(opts...))
```
Shouldn't it also include the above `panicHandler`?

srosenberg commented on pkg/cmd/roachtest/test_impl.go:
```diff
@@ -648,6 +653,22 @@ func (t *testImpl) IsBuildVersion(minVersion string) bool {
 
+func panicHandler(_ context.Context, name string, l *logger.Logger, r interface{}) error {
+	l.Printf("panic stack trace in task %s:\n%s", name, string(debug.Stack()))
+	return fmt.Errorf("panic (stack trace above): %v", r)
```
We probably want to match the error reporting to the existing handler in `test_runner.go`,

```
if r := recover(); r != nil && r != errTestFatal {
    // NB: we're careful to avoid t.Fatalf here, which re-panics.
    // Note that the error will be logged to a file, and the stack will
    // contain the source of the panic.
    t.Errorf("test panicked: %v", r)
}
```

Also shouldn't we check for the sentinel value (`errTestFatal`) since that's reserved for `t.Fatal`?

herkolategan replied:
> We probably want to match the error reporting to the existing handler in test_runner.go

Can do, although there's slight differences in the way panics are handled on the test goroutine level, vs. tasks. I'll update the wording on error to more closely match the above, but maybe just add the 'task' keyword in there so that we can discern that it was a task that panicked. The slight difference I'll explain below:

> Also shouldn't we check for the sentinel value (errTestFatal) since that's reserved for t.Fatal?

This is where the above logic comes into play when deciding to send a task event:
```
		case m.events <- event:
			// exit goroutine
		case <-m.ctx.Done():
			// Parent context already finished, exit goroutine.
```
`t.Fatal` cancels the context and reports an error to the test framework. The select logic ensures that the task's error does not get reported via `t.Errorf` because the context is canceled. This differs from how it's handled if a panic occurs in the main test goroutine, it requires the extra check for the sentinel value to ensure the error (panic) is not reported again as fatal has already reported it.

If we were to change the select logic above, event handling to report errors even if the context is cancelled we would need to handle the errors differently and check the panic sentinel error value etc.

herkolategan commented on pkg/cmd/roachtest/roachtestutil/task/manager.go:
```diff
@@ -0,0 +1,143 @@
+func (m *manager) defaultOptions() []Option {
+	// The default panic handler simply returns the panic as an error.
+	defaultPanicHandlerFn := func(_ context.Context, name string, l *logger.Logger, r interface{}) error {
+		return r.(error)
```
True, updated this to use format. The error and panic defaults actually do not get used (both mixedversion and the main test framework supply their own), but felt it necessary to have some minimal defaults at least.
